test/schema-special-usage.test.js

```diff
@@ -221,3 +222,66 @@ test('Should handle $patch keywords in body', t => {
     })
   })
 })
+
+test("serializer read validator's schemas", t => {
+  t.plan(4)
+  const ajvInstance = new AJV()
+
+  const baseSchema = {
+    $id: 'http://example.com/schemas/base',
+    definitions: {
+      hello: { type: 'string' }
+    },
+    type: 'object',
+    properties: {
+      hello: { $ref: '#/definitions/hello' }
+    }
+  }
+
+  const refSchema = {
+    $id: 'http://example.com/schemas/ref',
+    type: 'object',
+    properties: {
+      hello: { $ref: 'http://example.com/schemas/base#/definitions/hello' }
+    }
+  }
+
+  ajvInstance.addSchema(baseSchema)
+  ajvInstance.addSchema(refSchema)
+
+  const fastify = Fastify({
+    schemaController: {
+      bucket: function factory (storeInit) {
+        t.notOk(storeInit, 'is is always empty because fastify.addSchema is not called')
+        return {
+          hasNewSchemas () { return false },
+          getSchemas () {
+            return {
+              [baseSchema.$id]: ajvInstance.getSchema(baseSchema.$id).schema,
+              [refSchema.$id]: ajvInstance.getSchema(refSchema.$id).schema
+            }
+          }
+        }
+      }
+    }
+  })
+
+  fastify.setValidatorCompiler(function ({ schema }) {
+    return ajvInstance.compile(schema)
+  })
```

Comment:
@Eomm is this actually necessary? In my code I have the following test that does not return any error:

```js
tap.test('works with absolute references', async t => {
  const baseSchema = {
    $id: 'http://example.com/schemas/base',
    definitions: {
      hello: { type: 'string' }
    },
    type: 'object',
    properties: {
      hello: { $ref: '#/definitions/hello' }
    }
  };

  const refSchema = {
    $id: 'http://example.com/schemas/ref',
    type: 'object',
    properties: {
      hello: { $ref: 'http://example.com/schemas/base#/definitions/hello' }
    }
  };

  // This is my local container library.
  const ajvInstance = jsonSchema.withSchemas({ schemas: [baseSchema, refSchema] });
  const server = fastify({ schemaController: ajvInstance.fastifyBucket });

  server.route({
    path: '/',
    method: 'GET',
    schema: {
      response: {
        '2xx': ajvInstance.getSchema('http://example.com/schemas/ref').schema
      }
    },
    handler(req, res) {
      res.send({ hello: 'world' });
    }
  });

  const response = await server.inject({ url: '/' });
  t.same(response.json(), { hello: 'world' });
});
```

If a bucket is supplied, I do not think the user needs to define a validator compiler nor a schema compiler.

Reply:
> If a bucket is supplied, I do not think the user needs to define a validator compiler nor a schema compiler.

it is true; in this test, it seems odd to me to create an ajv instance to store the schemas without using it.
For example, the validation would be executed with the fastify's ajv setting and not with the user one.

I must add more tests for these usages 👍

Reply:
I do not understand.

Reply:
If the user doesn't set:

```
fastify.setValidatorCompiler(function ({ schema }) {
    return ajvInstance.compile(schema)
})
```

Fastify will instantiate and use an internal `ajv` instance to build the `validation` functions with the schemas returned by the `schemaController.bucket` of the user: so there would be an extra effort.

Reply:
🤔 okay. Given that explanation, I think I can devise a test for this tomorrow morning.